### PR TITLE
Make the slot as the SPI interface and add a @Order annotation for define slot order

### LINKED ISSUE #316: Slot chain SPI should target slot itself rather than the entire builder
## Issue Description

Type: *feature request*

### Describe what happened (or what feature you want)

Due to the fact that the order between slots in slot chain is essential, currently Sentinel make `SlotChainBuilder` as the slot chain SPI so that users can implement their own builder, where they can arrange the order between slots. But this introduce complexity for users to arrange the builder when there are many external slots from other dependency.

### Describe what you expected to happen

The ideal approach is to make the slot itself as the SPI interface so that users only need to care about their own slot, rather than the global chain. But due to the fact that the order between slots in slot chain is essential, we have to provide a `@Order` annotation for users so that they can arrange the order with the annotation. And the default slots of Sentinel (e.g. `StatisticSlot`) should have their own unique order respectively.

### PR DESCRIPTION
**Pull request by cdfive**
### Describe what this PR does / why we need it

Make it easier to add custorm slot and arrange its order for users.

### Does this pull request fix one issue?

Fixes #316 

### Describe how you did it

1.Add a `@Order` annotation and define the order of default slots from -9000 to -1000
NodeSelectorSlot: -9000
ClusterBuilderSlot: -8000
LogSlot: -7000
StatisticSlot: -6000
ParamFlowSlot: -5000
SystemSlot: -4000
AuthoritySlot: -3000
FlowSlot: -2000
DegradeSlot: -1000

2.Add a `loadOrderedInstanceList(Class<T>)` method in `SpiLoader`,
to Load the ordered instance list by SPI

3.Modify the `DefaultSlotChainBuilder` 
new XxxSlot changed to using `SpiLoader.loadOrderedInstanceList(ProcessorSlot.class);`

### Describe how to verify it

Add unit test

### Special notes for reviews

Please review its feasibility in this way, 
if it is okay, maybe the `InitOrder` can also use `Order` annotation instead, for similar logic,
the `InitOrder` and `InitExecutor` didn't modified at present.

### CODE REVIEW

**Comment by cdfive** [comment]
@sczyh30 Hi,
synchronize from master branch, I find `@SpiOrder`, so I use that instead of previous `@Order`.

Use `SpiLoader.loadInstanceListSorted(Class<T> clazz)` to get the sorted SPI instance list.

Since the ProcessorSlot need to create new instance in each `DefaultSlotChainBuilder`,
, so I modified the method, not use the `SERVICE_LOADER_MAP` but call ServiceLoader.load(clazz) directly, in order to get new instance each call.

**Comment by sczyh30** [comment]
Could you please fix the conflict?

**Comment by cdfive** [comment]
Sorry for late, the conflicts has fixed.

**Comment by sczyh30** [comment]
@cdfive Could you please resolve the conflict?

**Comment by cdfive** [comment]
> @cdfive Could you please resolve the conflict?
I'll do it these days.

**Comment by cdfive** [comment]
@sczyh30 Sorry for late, The conflicts has been resolved.
I found that the `ProcessorSlot`(such as `NodeSelectorSlot`,`ClusterBuilderSlot`...) should create new instance in `DefaultSlotChainBuilder`, so add a method `loadDifferentInstanceListSorted` in `SpiLoader` class.

`SpiLoaderTest#testLoadDifferentInstanceListSorted`,`DefaultSlotChainBuilderTest#testBuild`
to verify it, please check.

**Comment by sczyh30** [comment]
👍Great, I'll review the PR soon.

**Comment by cdfive** [comment]
Now `GatewaySlotChainBuilder`,`HotParamSlotChainBuilder` are removed, `DefaultSlotChainBuilder` is the only implement class of `SlotChainBuilder`,
which can sovle #1306, since the `SlotChainProvider#newSlotChain` will use `DefaultSlotChainBuilder`.

All slots are marked with `@SpiOrder`, from -10000 to -1000:
```
Slot                                @SpiOrder
NodeSelectorSlot           -10000
ClusterBuilderSlot            -9000
LogSlot                              -8000
StatisticSlot                       -7000
AuthoritySlot                     -6000
SystemSlot                        -5000
GatewayFlowSlot              -4000
ParamFlowSlot                  -3000
FlowSlot                             -2000
DegradeSlot                       -1000
```
Test cases and comment has been improved, please take a review.

Is there any other suggestion? @sczyh30 @jasonjoo2010 
Do we need to add the order values of Sentinel main slots into a Constant class, e.g. `Constans`?

like:
```
public static final String NODE_SELECTOR_SLOT_ORDER = -10000;
public static final String CLUSTER_BUILDER_SLOT_ORDER = NODE_SELECTOR_SLOT_ORDER + 1000;
...
public static final String SYSTEM_SLOT_ORDER = ...
```

and in extension `GatewayFlowSlot`:
```java
@SpiOrder(Constants.SYSTEM_SLOT_ORDER + 1000)
public class GatewayFlowSlot {
}
```
Maybe it will be useful for user and developer to see the orders of slots more intuitively.

**Comment by jasonjoo2010** [comment]
I think maybe  `GatewaySlotChainBuilder` and `HotParamSlotChainBuilder` should be preserved as legacy, declare them as deprecated, unlink their service declarations and rewrite them to inherit `DefaultSlotChainBuilder` to avoid that someone may reference or inherit them.

Thus we can allow users' customizations:

* Implement new ProcessorSlot by extends `AbstractProcessorSlot` and define it in their own META-INF
* Implement their own SlotChainBuilder and do all things in a different way (should be careful)

So **provider** of SlotChainBuilder maybe still need some adaption?

**Comment by cdfive** [comment]
`GatewaySlotChainBuilder` and `HotParamSlotChainBuilder` are reserved, marked as `@Deprecated`, the implement is not changed, and unlink their service declarations.
How about this? @jasonjoo2010 @sczyh30

File: sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/slot/GatewaySlotChainBuilder.java [hunk 1]
```diff
@@ -31,8 +31,13 @@
...
  * @author Eric Zhao
  * @since 1.6.1
+ *
+ * @deprecated since 1.7.2, we can use @SpiOrder(-4000) to adjust the order of slots,
+ * this class is reserved for compatibility with older versions.
+ * @see {@link GatewayFlowSlot}
  */
-public class GatewaySlotChainBuilder implements SlotChainBuilder {
+@Deprecated
+public class GatewaySlotChainBuilder extends DefaultSlotChainBuilder {
```
Comment (jasonjoo2010) [inline sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/slot/GatewaySlotChainBuilder.java]: Oh if you take my advice keeping an empty legacy one you should make it a real EMPTY one.
According to the new `DefaultSlotChainBuilder` i think there is no need keeping overriding of `build()` here, right?

File: sentinel-extension/sentinel-parameter-flow-control/src/main/java/com/alibaba/csp/sentinel/slots/HotParamSlotChainBuilder.java [hunk 2]
```diff
@@ -31,8 +31,13 @@
...
  * @author Eric Zhao
  * @since 0.2.0
+ *
+ * @deprecated since 1.7.2, we can use @SpiOrder(-3000) to adjust the order of slots,
+ * this class is reserved for compatibility with older versions.
+ * @see {@link ParamFlowSlot}
  */
-public class HotParamSlotChainBuilder implements SlotChainBuilder {
+@Deprecated
+public class HotParamSlotChainBuilder extends DefaultSlotChainBuilder {
```
Comment (jasonjoo2010) [inline sentinel-extension/sentinel-parameter-flow-control/src/main/java/com/alibaba/csp/sentinel/slots/HotParamSlotChainBuilder.java]: Same as GatewaySlotChainBuilder

**Comment by sczyh30** [comment]
> Okay i have no further question now.
> 
> @sczyh30 It's suggested that you make a quick review.

I'll take a look.

File: sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/slot/GatewaySlotChainBuilder.java (hunk 1 above)
Comment (cdfive) [reply sentinel-adapter/sentinel-api-gateway-adapter-common/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/common/slot/GatewaySlotChainBuilder.java]: I'm sorry for misunderstanding the point.It has been fixed, please check.

File: sentinel-extension/sentinel-parameter-flow-control/src/main/java/com/alibaba/csp/sentinel/slots/HotParamSlotChainBuilder.java (hunk 2 above)
Comment (cdfive) [reply sentinel-extension/sentinel-parameter-flow-control/src/main/java/com/alibaba/csp/sentinel/slots/HotParamSlotChainBuilder.java]: Fixed, please check.

**Comment by sczyh30** [comment]
Nice work. Thanks!

**Comment by sczyh30** [comment]
Here we need to replace the old placeholder with "{}":
https://github.com/alibaba/Sentinel/blob/master/sentinel-core/src/main/java/com/alibaba/csp/sentinel/util/SpiLoader.java#L222

**Comment by sczyh30** [comment]
And we may need to improve the logs here because it will yield a large number of logs every time a  new chain is built:

```
2020-03-26 16:21:14.948 INFO Add child </metric/queryTopResourceMetric.json> to node <sentinel_web_servlet_context>
2020-03-26 16:21:36.215 INFO [SpiLoader] Found {0} SPI: {1} with order -10000
2020-03-26 16:21:36.216 INFO [SpiLoader] Found {0} SPI: {1} with order -9000
2020-03-26 16:21:36.217 INFO [SpiLoader] Found {0} SPI: {1} with order -8000
2020-03-26 16:21:36.217 INFO [SpiLoader] Found {0} SPI: {1} with order -7000
2020-03-26 16:21:36.217 INFO [SpiLoader] Found {0} SPI: {1} with order -5000
2020-03-26 16:21:36.217 INFO [SpiLoader] Found {0} SPI: {1} with order -6000
2020-03-26 16:21:36.217 INFO [SpiLoader] Found {0} SPI: {1} with order -2000
2020-03-26 16:21:36.218 INFO [SpiLoader] Found {0} SPI: {1} with order -1000
2020-03-26 16:21:36.218 INFO [SpiLoader] Found {0} SPI: {1} with order -3000
2020-03-26 16:21:36.219 INFO [SpiLoader] Found {0} SPI: {1} with order -4000
2020-03-26 16:21:36.219 INFO Add child </resource/machineResource.json> to node <sentinel_web_servlet_context>
2020-03-26 16:21:38.956 INFO [SpiLoader] Found {0} SPI: {1} with order -10000
2020-03-26 16:21:38.956 INFO [SpiLoader] Found {0} SPI: {1} with order -9000
2020-03-26 16:21:38.956 INFO [SpiLoader] Found {0} SPI: {1} with order -8000
2020-03-26 16:21:38.957 INFO [SpiLoader] Found {0} SPI: {1} with order -7000
2020-03-26 16:21:38.957 INFO [SpiLoader] Found {0} SPI: {1} with order -5000
2020-03-26 16:21:38.957 INFO [SpiLoader] Found {0} SPI: {1} with order -6000
2020-03-26 16:21:38.957 INFO [SpiLoader] Found {0} SPI: {1} with order -2000
2020-03-26 16:21:38.957 INFO [SpiLoader] Found {0} SPI: {1} with order -1000
2020-03-26 16:21:38.958 INFO [SpiLoader] Found {0} SPI: {1} with order -3000
2020-03-26 16:21:38.958 INFO [SpiLoader] Found {0} SPI: {1} with order -4000
```

**Comment by cdfive** [comment]
> Here we need to replace the old placeholder with "{}":
> https://github.com/alibaba/Sentinel/blob/master/sentinel-core/src/main/java/com/alibaba/csp/sentinel/util/SpiLoader.java#L222

I'm sorry not merge master before commit. May I post new PR to fix it?

**Comment by cdfive** [comment]
> LGTM. The name `loadDifferentInstanceListSorted` might be confusing, maybe we could improve it later :)

How about named the method `loadPrototypeInstanceListSorted`? In `@scope` of Spring,
there are `singleton` and `prototype` for singleton or muti instances.

> And we may need to improve the logs here because it will yield a large number of logs every time a new chain is built:

Yes, since every new chain need create new slot instance, here is not suitable for too many logs.
How about using `RecordLog.debug` instead of `RecordLog.info` here? Now I couldn't thought out a better way,is there any suggestions? @sczyh30 @jasonjoo2010

**Comment by jasonjoo2010** [comment]
> > LGTM. The name `loadDifferentInstanceListSorted` might be confusing, maybe we could improve it later :)
> 
> How about naming the method `loadPrototypeInstanceListSorted`? In `@scope` of Spring,
> there are `singleton` and `prototype` for singleton or multiple instances.
> 
> > And we may need to improve the logs here because it will yield a large number of logs every time a new chain is built:
> 
> Yes, since every new chain need create new slot instance, here is not suitable for too many logs.
> How about using `RecordLog.debug` instead of `RecordLog.info` here? Now I couldn't thought out a better way,is there any suggestions? @sczyh30 @jasonjoo2010

I think it can be in trace or debug level

**Comment by jasonjoo2010** [comment]
And coming to the logging please use placeholders `{}` instead of joining of strings

**Comment by jasonjoo2010** [comment]
> And coming to the logging please use placeholders `{}` instead of joining of strings

... and i noticed you mentioned about new PR. Have you ever used `rebase` ? You can rebase your branch with an up to date master. It's suggested that making a backup before that. eg. 

```
git checkout master
git pull github master     <<<<< make master branch up to date
git checkout slot-order   <<<<< backup it to another temporary branch first
git rebase master   <<<< maybe you should solve some conflicts
```
Where `github` is a remote name of `alibaba/Sentinel`

**Comment by cdfive** [comment]
@jasonjoo2010 Thanks for detailed guideline! Previously I just use rebase in local, push -f to my forked repository and then post a PR in the UI of github site. Now the previous PR has been merged, could use rebase master? I'm worry about the safey.

**Comment by jasonjoo2010** [comment]
> @jasonjoo2010 Thanks for detailed guideline! Previously I just use rebase in local, push -f to my forked repository and then post a PR in the UI of github site. Now the previous PR has been merged, could use rebase master? I'm worry about the safey.

Yes rebase in your own repo. Is it dangerous? But it has been merged so the only thing you can do is to remove it :)

More polishings should lead to a new PR as you mentioned.

**Comment by cdfive** [comment]
Very clear! Got it, thanks! @jasonjoo2010

**Pull request merged**